File: scripts/differential/java/P7T3.java
package app.freerouting.autoroute.maze;

import app.freerouting.autoroute.maze.AutorouteControl.ExpansionCostFactor;
import app.freerouting.autoroute.AutorouteAttemptResult;
import app.freerouting.board.facade.RoutingBoard;
import app.freerouting.board.model.items.Item;
import app.freerouting.board.model.items.Pin;
import app.freerouting.board.model.items.Via;
import app.freerouting.board.model.structure.AngleRestriction;
import app.freerouting.board.trace.PolylineTrace;
import app.freerouting.geometry.planar.FloatPoint;
import app.freerouting.geometry.planar.IntPoint;
import app.freerouting.geometry.planar.Line;
import app.freerouting.geometry.planar.Point;
import app.freerouting.geometry.planar.Polyline;
import app.freerouting.settings.RouterSettings;
import app.freerouting.settings.sources.DefaultSettings;
import java.io.FileDescriptor;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.SortedSet;
import java.util.TreeSet;

/**
 * Plan 7 Task 5 differential driver, board level: {@code RoutingBoard.optChangedArea}
 * (RoutingBoard.java:151-161 -> :171-190 -> RoutingBoardOperations.java:52-79) and the {@code
 * TraceTightener.optChangedArea(ExpansionCostFactor[])} sweep it drives
 * (TraceTightener.java:121-169), over a real DSN board whose changed area was marked by real
 * routing.
 *
 * <p>Usage: {@code P7T3 <dsn> [mode] [accuracy] [routeK]}. Defaults: {@code mode = 0}, {@code
 * accuracy = 500}, {@code routeK = 6}.
 *
 * <p>The first {@code routeK} connections are routed with {@code P6T1}'s own machinery — its
 * {@code loadBoard} / {@code pickConnections} / {@code route} are package-private statics, which
 * is why this driver declares {@code package app.freerouting.autoroute.maze} rather than the
 * brief's {@code board.optimize}, and why {@code P6T1.java} is compiled alongside it (the {@code
 * P7T7} / {@code P7T10} precedent: the two drivers cannot then describe different boards). {@code
 * AutoroutePassRunner.java:224} calls {@code startMarkingChangedArea()} before every connection
 * and {@code P6T1.route} reproduces that, so after the loop {@code board.changedArea} holds
 * exactly the regions the router touched — which is the input {@code optChangedArea} is defined
 * over.
 *
 * <p>Modes, which change only what happens <b>after</b> the routing, so every mode sweeps the same
 * routed board:
 *
 * <ul>
 *   <li>{@code 0} — no vias offered to the optimiser ({@code traceCosts = null}) and the angle
 *       restriction forced to {@code NINETY_DEGREE};
 *   <li>{@code 1} — the same, {@code FORTYFIVE_DEGREE};
 *   <li>{@code 2} — the same, {@code NONE} (any angle);
 *   <li>{@code 3} — mode 2 plus {@code pinEdgeToTurnDist = 500}, which opens {@code
 *       PolylineTrace.pullTight:841-861} and therefore the {@code swap}/{@code
 *       correctConnectionToPin} pair inside the sweep;
 *   <li>{@code 4} — mode 2 plus a non-null {@code traceCosts}, which opens the {@code
 *       ViaOptimizer.optViaLocation} arm at {@code TraceTightener.java:160-165}. <b>This mode is
 *       expected to diff until Plan 7 Task 6 lands {@code ViaOptimizer}</b>; the port stubs the
 *       arm to {@code false} behind an {@code obligation:} marker naming that task.
 * </ul>
 *
 * <p><b>The budget is disabled on both sides.</b> Java's callers all pass the literal {@code
 * TIME_LIMIT_TO_PREVENT_ENDLESS_LOOP = 1000} into {@code optChangedArea}'s {@code timeLimit}
 * parameter, and {@code TraceTightener}'s constructor builds a {@code TimeLimit} only when {@code
 * timeLimit > 0} (TraceTightener.java:73-77). At <b>this</b> entry point the limit is therefore a
 * plain argument, not a constant to be patched: the driver passes {@code 0}, which is Java's own
 * "no limit", and the port passes {@code RouterBudget::disabled()}, whose {@code
 * opt_changed_area_ms} is {@code 0} for the same reason. No reflection is needed or used, and a
 * wall-clock difference between the two languages therefore cannot move the result.
 *
 * <p>The output is the {@code HEADER} line ({@code P6T1}'s convention: a run against the wrong jar
 * is a diff, not a silent pass), one line per routed connection's outcome, then the {@code
 * changedArea} per layer before the sweep, then the whole board in {@code getItems()} order —
 * descending id, quirk #63 — in {@code P6T15aProbe}'s polyline format.
 */
public final class P7T3 {

  private P7T3() {}

  static PrintStream out;

  public static void main(String[] args) throws Exception {
    if (args.length < 1) {
      System.err.println("usage: P7T3 <dsn> [mode] [accuracy] [routeK]");
      System.exit(2);
    }
    out = new PrintStream(new FileOutputStream(FileDescriptor.out), true, StandardCharsets.UTF_8);
    System.setOut(new PrintStream(OutputStream.nullOutputStream()));

    Path dsn = Paths.get(args[0]).toAbsolutePath().normalize();
    int mode = args.length > 1 ? Integer.parseInt(args[1]) : 0;
    int accuracy = args.length > 2 ? Integer.parseInt(args[2]) : 500;
    int routeK = args.length > 3 ? Integer.parseInt(args[3]) : 6;

    Path jar =
        Paths.get(RoutingBoard.class.getProtectionDomain().getCodeSource().getLocation().toURI())
            .toRealPath();
    out.printf(
        "HEADER jar=%s bytes=%d mtime=%d fixture=%s mode=%d accuracy=%d routeK=%d%n",
        jar,
        Files.size(jar),
        Files.getLastModifiedTime(jar).toMillis(),
        dsn.getFileName(),
        mode,
        accuracy,
        routeK);
    System.err.println("java-version " + System.getProperty("java.version"));

    RoutingBoard board = P6T1.loadBoard(dsn, null);
    RouterSettings settings = new DefaultSettings().getSettings();
    settings.setLayerCount(board.getLayerCount());
    settings.applyBoardSpecificOptimizations(board);

    // `P6T1.route` rather than `P6T1.routeOne`: the per-connection JSON is `p6t1`'s comparison
    // surface and duplicating its renderer in this twin would buy nothing. What this driver needs
    // from the routing is the *board* and its `changedArea`, so it prints only the outcome.
    for (P6T1.Connection connection : P6T1.pickConnections(board, routeK)) {
      Item item = board.getItem(connection.itemId());
      if (item == null) {
        out.println("route k=" + connection.k() + " item=" + connection.itemId() + " state=GONE");
        continue;
      }
      board.startMarkingChangedArea();
      SortedSet<Item> rippedItemList = new TreeSet<>();
      Map<Item, Integer> ripupCosts = new LinkedHashMap<>();
      AutorouteAttemptResult result =
          P6T1.route(
              board, settings, item, connection.netNo(), rippedItemList, ripupCosts, 1);
      out.println(
          "route k="
              + connection.k()
              + " item="
              + connection.itemId()
              + " net="
              + connection.netNo()
              + " state="
              + result.state
              + " ripped="
              + rippedItemList.size());
    }

    // The regime and the pin-edge distance are set **after** the routing, so all five modes sweep
    // the same board.
    AngleRestriction regime =
        switch (mode) {
          case 0 -> AngleRestriction.NINETY_DEGREE;
          case 1 -> AngleRestriction.FORTYFIVE_DEGREE;
          default -> AngleRestriction.NONE;
        };
    board.rules.setTraceAngleRestriction(regime);
    if (mode == 3) {
      // The corpus DSNs already carry a positive `pinEdgeToTurnDist` (rpi-splitter's is 20320),
      // so `PolylineTrace.pullTight:841-861` is open in modes 1 and 2 as well. Mode 3 widens it
      // to five times that, which moves the `checkConnectionToPin:1074-1076` threshold past every
      // routed stub and turns the pair from occasionally-firing into routinely-firing.
      board.rules.setPinEdgeToTurnDist(100000);
    }
    ExpansionCostFactor[] traceCosts = null;
    if (mode == 4) {
      traceCosts = new ExpansionCostFactor[board.getLayerCount()];
      for (int i = 0; i < traceCosts.length; i++) {
        traceCosts[i] = new ExpansionCostFactor(1.0, 1.0);
      }
    }

    out.println(
        "sweep regime="
            + regime
            + " pinEdgeToTurnDist="
            + Double.toString(board.rules.getPinEdgeToTurnDist())
            + " traceCosts="
            + (traceCosts == null ? "null" : traceCosts.length));
    dumpChangedArea(board, "before");

    // `clipShape = null` runs `RoutingBoardOperations.java:64`'s branch — the reference comparison
    // ruling 9 is about. `stoppableThread = null` and `timeLimit = 0` are the disabled budget.
    board.optChangedArea(new int[0], null, accuracy, traceCosts, null, 0);

    dumpChangedArea(board, "after");
    dumpBoard(board);
  }

  static void dumpChangedArea(RoutingBoard board, String tag) {
    if (board.changedArea == null) {
      out.println("changedArea " + tag + "=null");
      return;
    }
    StringBuilder sb = new StringBuilder("changedArea " + tag + "=[");
    for (int i = 0; i < board.getLayerCount(); i++) {
      if (i > 0) {
        sb.append(",");
      }
      var area = board.changedArea.getArea(i);
      // `IntOctagon.toString()` is `Object`'s, i.e. the identity hash — which `-XX:hashCode=2`
      // pins to a constant and the port cannot reproduce. Print the eight coordinates instead;
      // `isEmpty()` is Java's reference test against the `EMPTY` singleton, which
      // `ChangedArea.getArea:107` returns for an untouched layer.
      sb.append(
          area.isEmpty()
              ? "empty"
              : "("
                  + area.leftX
                  + ","
                  + area.bottomY
                  + ","
                  + area.rightX
                  + ","
                  + area.topY
                  + ","
                  + area.upperLeftDiagonalX
                  + ","
                  + area.lowerRightDiagonalX
                  + ","
                  + area.lowerLeftDiagonalX
                  + ","
                  + area.upperRightDiagonalX
                  + ")");
    }
    out.println(sb.append("]"));
  }

  // --- dumps ---------------------------------------------------------------------------------

  static String ln(Line l) {
    return "("
        + ((IntPoint) l.a).x
        + ","
        + ((IntPoint) l.a).y
        + ")->("
        + ((IntPoint) l.b).x
        + ","
        + ((IntPoint) l.b).y
        + ")";
  }

  static String pt(Polyline p, int no) {
    Point corner = p.corner(no);
    if (corner instanceof IntPoint ip) {
      return "(" + ip.x + "," + ip.y + ")";
    }
    FloatPoint f = p.cornerApprox(no);
    return "~(" + Double.toString(f.x) + "," + Double.toString(f.y) + ")";
  }

  static String poly(Polyline p) {
    if (p == null) {
      return "null";
    }
    StringBuilder sb = new StringBuilder();
    sb.append("n=").append(p.lines.length).append(" lines=[");
    for (int i = 0; i < p.lines.length; i++) {
      if (i > 0) {
        sb.append(",");
      }
      sb.append(ln(p.lines[i]));
    }
    sb.append("] corners=[");
    for (int i = 0; i < p.cornerCount(); i++) {
      if (i > 0) {
        sb.append(",");
      }
      sb.append(pt(p, i));
    }
    return sb.append("]").toString();
  }

  static String pointOf(Point p) {
    FloatPoint f = p.toFloat().round().toFloat();
    return "(" + (int) f.x + "," + (int) f.y + ")";
  }

  static String nets(int[] netNos) {
    StringBuilder sb = new StringBuilder("[");
    for (int i = 0; i < netNos.length; i++) {
      if (i > 0) {
        sb.append(",");
      }
      sb.append(netNos[i]);
    }
    return sb.append("]").toString();
  }

  static void dumpBoard(RoutingBoard board) {
    out.println("maxId=" + board.communication.idGenerator.maxGeneratedId());
    for (Item item : board.getItems()) {
      StringBuilder sb = new StringBuilder();
      sb.append("item id=")
          .append(item.getId())
          .append(" type=")
          .append(item.getClass().getSimpleName())
          .append(" nets=")
          .append(nets(item.netNumbers))
          .append(" cl=")
          .append(item.clearanceClassIndex())
          .append(" fix=")
          .append(item.getFixedState());
      if (item instanceof PolylineTrace trace) {
        sb.append(" layer=")
            .append(trace.getLayer())
            .append(" hw=")
            .append(trace.getHalfWidth())
            .append(" ")
            .append(poly(trace.polyline()));
      } else if (item instanceof Via via) {
        sb.append(" center=").append(pointOf(via.getCenter()));
      } else if (item instanceof Pin pin) {
        sb.append(" center=").append(pointOf(pin.getCenter()));
      }
      out.println(sb);
    }
  }
}

File: scripts/differential/rust/src/bin/p7t3.rs
//! Rust twin of `scripts/differential/java/P7T3.java` (Plan 7 Task 5): `RoutingBoard.optChangedArea`
//! (RoutingBoard.java:151-161 -> :171-190 -> RoutingBoardOperations.java:52-79) and the
//! `TraceTightener.optChangedArea(ExpansionCostFactor[])` sweep it drives (TraceTightener.java:
//! 121-169), over a real DSN board whose changed area was marked by real routing.
//!
//! Usage: `p7t3 <dsn> [mode] [accuracy] [routeK]`. See the Java twin's class comment for the five
//! modes and for why mode 4 is expected to diff until Plan 7 Task 6 lands `ViaOptimizer`.
//!
//! **The budget is disabled on both sides.** `RouterBudget::disabled()`'s `opt_changed_area_ms` is
//! `0`, which is what the Java half passes for `timeLimit`, and `TraceTightener`'s constructor
//! builds a `TimeLimit` only when `timeLimit > 0` (TraceTightener.java:73-77). So neither side
//! reads a clock and a wall-clock difference between the two languages cannot move the result.
//!
//! `P6T1.java`'s four routing choices are transcribed here for the reason `p7t7.rs`'s module
//! comment gives: Rust binaries cannot share a private module, and the Java side is the
//! un-duplicated one, so drift shows up as a diff.

use std::collections::{BTreeMap, BTreeSet};
use std::io::{BufWriter, Write};
use std::time::UNIX_EPOCH;

use fr_board::items::Item;
use fr_board::prelude::*;
use fr_dsn::parser::scope_parameter::DsnReadOptions;
use fr_dsn::{java_double_to_string, BoardReadResult};
use fr_geometry::{Line, Point, Polyline};
use fr_router::board_ext::RoutingBoardExt;
use fr_router::pipeline::RouterBudget;
use fr_router::route_connection;
use fr_settings::sources::DefaultSettings;
use fr_settings::{ExpansionCostFactor, HostEnvironment, RouterSettings, SettingsSource};

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    if args.is_empty() {
        eprintln!("usage: p7t3 <dsn> [mode] [accuracy] [routeK]");
        std::process::exit(2);
    }
    let dsn = std::fs::canonicalize(&args[0])
        .unwrap_or_else(|e| panic!("cannot resolve {}: {e}", args[0]));
    let mode: i32 = args.get(1).map_or(0, |a| a.parse().expect("mode"));
    let accuracy: i32 = args.get(2).map_or(500, |a| a.parse().expect("accuracy"));
    let route_k: usize = args.get(3).map_or(6, |a| a.parse().expect("routeK"));

    let stdout = std::io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    print_header(&mut out, &dsn, mode, accuracy, route_k);

    let mut board = load_board(&dsn);
    let settings = build_settings(&board);

    for connection in pick_connections(&board, route_k) {
        route_one(&mut out, &mut board, &settings, &connection);
    }

    let regime = match mode {
        0 => AngleRestriction::NinetyDegree,
        1 => AngleRestriction::FortyFiveDegree,
        _ => AngleRestriction::None,
    };
    board.rules.trace_angle_restriction = regime;
    if mode == 3 {
        board.rules.set_pin_edge_to_turn_dist(100_000.0);
    }
    let trace_costs: Option<Vec<ExpansionCostFactor>> = if mode == 4 {
        Some(vec![
            ExpansionCostFactor {
                horizontal: 1.0,
                vertical: 1.0,
            };
            board.get_layer_count()
        ])
    } else {
        None
    };

    writeln!(
        out,
        "sweep regime={} pinEdgeToTurnDist={} traceCosts={}",
        regime_name(regime),
        java_double_to_string(board.rules.get_pin_edge_to_turn_dist()),
        match &trace_costs {
            Some(costs) => costs.len().to_string(),
            None => "null".to_string(),
        }
    )
    .expect("write");
    dump_changed_area(&mut out, &board, "before");

    // `clip_shape = None` runs `RoutingBoardOperations.java:64`'s branch — the reference
    // comparison ruling 9 is about. The stop check never trips and the budget is disabled.
    let budget = RouterBudget::disabled();
    board
        .opt_changed_area(
            None,
            &[],
            None,
            accuracy,
            trace_costs.as_deref(),
            &|| false,
            budget.opt_changed_area_ms,
        )
        .expect("optChangedArea cannot fail in Java");

    dump_changed_area(&mut out, &board, "after");
    dump_board(&mut out, &board);
    out.flush().expect("flush");
}

// ------------------------------------------------------------------------------------------------
// Header, board, settings, routing — `P6T1.java`'s choices, transcribed
// ------------------------------------------------------------------------------------------------

fn print_header<W: Write>(
    out: &mut W,
    dsn: &std::path::Path,
    mode: i32,
    accuracy: i32,
    route_k: usize,
) {
    let jar =
        std::env::var("FREEROUTING_JAR").expect("environment variable FREEROUTING_JAR is not set");
    let jar = std::fs::canonicalize(jar).expect("jar exists");
    let meta = std::fs::metadata(&jar).expect("jar metadata");
    let mtime = meta
        .modified()
        .expect("mtime")
        .duration_since(UNIX_EPOCH)
        .expect("after epoch")
        .as_millis();
    writeln!(
        out,
        "HEADER jar={} bytes={} mtime={mtime} fixture={} mode={mode} accuracy={accuracy} routeK={route_k}",
        jar.display(),
        meta.len(),
        dsn.file_name().expect("a file name").to_string_lossy(),
    )
    .expect("write");
    if let Ok(exe) = std::env::current_exe() {
        eprintln!("rust-binary {}", exe.display());
    }
}

fn load_board(dsn: &std::path::Path) -> Board {
    let file = std::fs::File::open(dsn).unwrap_or_else(|e| panic!("cannot open {dsn:?}: {e}"));
    let design_name = dsn
        .file_name()
        .expect("a file name")
        .to_string_lossy()
        .into_owned();
    let result = fr_dsn::read_board(file, None, Some(&design_name), &DsnReadOptions::default());
    match result {
        BoardReadResult::Success { board, .. } | BoardReadResult::OutlineMissing { board, .. } => {
            *board.unwrap_or_else(|| panic!("{design_name} produced no board"))
        }
        other => panic!("{design_name} did not read: {other:?}"),
    }
}

fn build_settings(board: &Board) -> RouterSettings {
    let host = HostEnvironment::detect();
    let mut settings = DefaultSettings::new(&host)
        .get_settings()
        .expect("DefaultSettings always answers a table")
        .clone();
    settings.set_layer_count(board.get_layer_count());
    settings.apply_board_specific_optimizations(board);
    settings
}

struct Connection {
    k: usize,
    item_id: ItemId,
    net_no: i32,
}

fn pick_connections(board: &Board, max_items: usize) -> Vec<Connection> {
    let mut result = Vec::new();
    if max_items == 0 {
        return result;
    }
    let mut k = 0;
    for item_id in board.items_in_board_order() {
        let Some(item) = board.get_item(item_id) else {
            continue;
        };
        if item.as_connectable().is_none() {
            continue;
        }
        let net_nos: Vec<i32> = item.net_nos().to_vec();
        for net_no in net_nos {
            if board.unconnected_set(item_id, net_no).is_empty() {
                continue;
            }
            k += 1;
            result.push(Connection {
                k,
                item_id,
                net_no,
            });
            if result.len() >= max_items {
                return result;
            }
        }
    }
    result
}

fn route_one<W: Write>(
    out: &mut W,
    board: &mut Board,
    settings: &RouterSettings,
    connection: &Connection,
) {
    if board.get_item(connection.item_id).is_none() {
        writeln!(
            out,
            "route k={} item={} state=GONE",
            connection.k, connection.item_id.0
        )
        .expect("write");
        return;
    }
    board.start_marking_changed_area();
    let mut ripped: BTreeSet<ItemId> = BTreeSet::new();
    let mut ripup_costs: BTreeMap<ItemId, i32> = BTreeMap::new();
    let trace_costs = settings.get_trace_costs();
    let mut engine = None;
    let result = route_connection(
        board,
        &mut engine,
        connection.item_id,
        connection.net_no,
        settings,
        &trace_costs,
        &mut ripped,
        &mut ripup_costs,
        1,
        settings.get_start_ripup_costs(),
        !settings.is_fanout_enabled(),
        false,
        &|| false,
    );
    writeln!(
        out,
        "route k={} item={} net={} state={} ripped={}",
        connection.k,
        connection.item_id.0,
        connection.net_no,
        result.state.name(),
        ripped.len()
    )
    .expect("write");
}

// ------------------------------------------------------------------------------------------------
// Dumps — `P7T3.java`'s
// ------------------------------------------------------------------------------------------------

fn regime_name(angle: AngleRestriction) -> &'static str {
    match angle {
        AngleRestriction::NinetyDegree => "NINETY_DEGREE",
        AngleRestriction::FortyFiveDegree => "FORTYFIVE_DEGREE",
        AngleRestriction::None => "NONE",
    }
}

fn dump_changed_area<W: Write>(out: &mut W, board: &Board, tag: &str) {
    let Some(changed_area) = &board.changed_area else {
        writeln!(out, "changedArea {tag}=null").expect("write");
        return;
    };
    let mut parts = Vec::new();
    for i in 0..board.get_layer_count() {
        let area = changed_area.get_area(i);
        parts.push(if area.is_empty() {
            "empty".to_string()
        } else {
            format!(
                "({},{},{},{},{},{},{},{})",
                area.left_x,
                area.bottom_y,
                area.right_x,
                area.top_y,
                area.upper_left_diagonal_x,
                area.lower_right_diagonal_x,
                area.lower_left_diagonal_x,
                area.upper_right_diagonal_x
            )
        });
    }
    writeln!(out, "changedArea {tag}=[{}]", parts.join(",")).expect("write");
}

fn dump_line(line: &Line) -> String {
    format!("({},{})->({},{})", line.a.x, line.a.y, line.b.x, line.b.y)
}

fn dump_corner(polyline: &Polyline, no: usize) -> String {
    match polyline.corner(no).expect("no is below cornerCount") {
        Point::Int(p) => format!("({},{})", p.x, p.y),
        Point::Rational(_) => {
            let f = polyline.corner_approx(no).expect("no is below cornerCount");
            format!(
                "~({},{})",
                java_double_to_string(f.x),
                java_double_to_string(f.y)
            )
        }
    }
}

fn dump_polyline(polyline: &Polyline) -> String {
    let lines: Vec<String> = polyline.lines().iter().map(dump_line).collect();
    let corners: Vec<String> = (0..polyline.corner_count())
        .map(|i| dump_corner(polyline, i))
        .collect();
    format!(
        "n={} lines=[{}] corners=[{}]",
        polyline.lines().len(),
        lines.join(","),
        corners.join(",")
    )
}

fn dump_point(point: &Point) -> String {
    let rounded = point.to_float().round();
    format!("({},{})", rounded.x, rounded.y)
}

fn dump_nets(net_nos: &[i32]) -> String {
    let inner: Vec<String> = net_nos.iter().map(i32::to_string).collect();
    format!("[{}]", inner.join(","))
}

fn dump_fixed_state(state: FixedState) -> &'static str {
    match state {
        FixedState::Unfixed => "UNFIXED",
        FixedState::ShoveFixed => "SHOVE_FIXED",
        FixedState::UserFixed => "USER_FIXED",
        FixedState::SystemFixed => "SYSTEM_FIXED",
    }
}

fn dump_board<W: Write>(out: &mut W, board: &Board) {
    writeln!(
        out,
        "maxId={}",
        board.communication.id_gen.max_generated_id()
    )
    .expect("write");
    for item in board.get_items() {
        let type_name = match item {
            Item::Trace(_) => "PolylineTrace",
            Item::Via(_) => "Via",
            Item::Pin(_) => "Pin",
            Item::ObstacleArea(_) => "ObstacleArea",
            Item::ConductionArea(_) => "ConductionArea",
            Item::ViaObstacleArea(_) => "ViaObstacleArea",
            Item::ComponentObstacleArea(_) => "ComponentObstacleArea",
            Item::ComponentOutline(_) => "ComponentOutline",
            Item::BoardOutline(_) => "BoardOutline",
        };
        let mut line = format!(
            "item id={} type={} nets={} cl={} fix={}",
            item.id().0,
            type_name,
            dump_nets(item.net_nos()),
            item.clearance_class(),
            dump_fixed_state(item.get_fixed_state())
        );
        match item {
            Item::Trace(trace) => {
                line.push_str(&format!(
                    " layer={} hw={} {}",
                    trace.get_layer(),
                    trace.get_half_width(),
                    dump_polyline(trace.polyline())
                ));
            }
            Item::Via(_) | Item::Pin(_) => {
                let center = board
                    .drill_center(item.id())
                    .expect("a drill item has a centre");
                line.push_str(&format!(" center={}", dump_point(&center)));
            }
            _ => {}
        }
        writeln!(out, "{line}").expect("write");
    }
}
